takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/compile/AbstractCompileMojo.java

```diff
@@ -419,6 +431,9 @@ public void execute() throws MojoExecutionException, MojoFailureException {
         log.info("Compiling {} sources to {}", sources.size(), getOutputDirectory());
         int compiled = compiler.compile();
         log.info("Compiled {} out of {} sources ({} ms)", compiled, sources.size(), stopwatch.elapsed(TimeUnit.MILLISECONDS));
+        if (unusedDeclaredDependency != AccessRulesViolation.ignore && context.isEscalated()) {
```

Comment:
True. I did not consider that scenario, and I agree with what you said. I'll remove this change.

I do not see another way from what you detailed. Since there is a need of persistence of the set of dependencies referenced between builds, I believe to keep that set of dependencies accurate, each source's list of dependencies would need to be tracked.
Along those lines, I'm wondering if something like a `Map<Path, Set<Artifact>>` where at the end of each build, any sources that were built would update the map would work (or the reverse with each dependency listing each source that references it where any that are still empty by the end of the build are considered unused).